### ISSUE TITLE
Ability to move inputs/parameters between memory locations (e.g. host, device)

### ISSUE THREAD

**Issue by nvukobratTT** [tenstorrent/tt-forge-fe#175]
### Summary
For more flexible inference/training we need more granular control over:
- Where specific tensors should live
- Where specific modules/graphs should execute

The proposal is to use the `.to("cpu"/"tt")` functionality (similar to torch 2.0).

### Details
In order to:
- Define should inputs be located on the host/device
- Define where the specific model should live/execute (host/CPU or tt device)

It's adequate to have API to support this kind of control. Therefore, the proposal is to use a style inspired by Torch 2.0. 

For example:
```py
    # Put inputs on device
    input_tensor = torch.rand(1, 784).to("tt")

    # Define model and instruct it to compile and run on TT device
    framework_model = MNISTLinear()
    framework_model.to("tt")

    loss = torch.nn.L1Loss().to("cpu")

    # Compiles the model
    tt_model = pybuda.compile(framework_model)
    cpu_model = pybuda.compile(loss)
```

We should investigate can we utilize metadata from torch elements. E.g. is it specified to be run on CPU (`to("cpu")`) or some other backed. If it's possible, we can follow above example and use `pybuda.compile` for CPU devices as well. This can nicely integrate in a flow where we need to be aware of specific inputs/ouputs for CPU models as well (e.g. loss or optimizer run on CPU; we need to tell runtime from where to gather appropriate tensors).

### Passing criteria
- We have flexibility to define where inputs will be stored before model run
- We have flexibility to define where model will be executed (compiled)
- `pybuda.compile` should return to the runtime appropriate tensor layout details (e.g. is tensor stored on host, dram, l1, etc.)

**Comment by staylorTT** [comment]
@vladimirjovanovicTT Is this issue still updated with the latest status? 

e.g. Screen, Open, etc?

**Comment by vladimirjovanovicTT** [comment]
It's a bit of an old issue but this will be solved when we implement runtime stitching support through FFE

**Comment by nvukobratTT** [comment]
@vladimirjovanovicTT I think this one is more related to the API of pushing and pulling weights to and from the device, rather than support for runtime stitching. 

Runtime stitching potential dependency, but isn't the actual feature. This one requires a bit more API magic to make it functional :))
